src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs

```diff
@@ -60,30 +69,86 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
-            var authenticator = CreateAuthenticator(description);
-            authenticator.Authenticate(connection, description, cancellationToken);
+            // if we've already piggybacked the initial isMaster request, we'll already have SaslSupportedMechs
+            // as part of the response. This check can be removed once multiple authenticators are no longer supported.
+            // Additionally, we should only attempt to saslSupportedMechs negotiation on servers >=4.0. 
+            if (description.IsMasterResult.SaslSupportedMechsReceived 
+                || !Feature.ScramSha256Authentication.IsSupported(description.ServerVersion)) 
+            {
+                CreateAuthenticator(connection, description).Authenticate(connection, description, cancellationToken);
+                return;
+            } // else, we're in a multi-authenticator--we did not piggyback, so we have to query for saslSupportedMechs 
+            
+            var command = CustomizeInitialIsMasterCommand(IsMasterHelper.CreateCommand());
+            var isMasterProtocol = IsMasterHelper.CreateProtocol(command);
+            var isMasterResult = IsMasterHelper.GetResult(connection, isMasterProtocol, cancellationToken);
+            var mergedIsMasterResult = new IsMasterResult(description.IsMasterResult.Wrapped.Merge(isMasterResult.Wrapped));
+            var mergedDescription = new ConnectionDescription(
+                description.ConnectionId, 
+                mergedIsMasterResult, 
+                description.BuildInfoResult);
+
+            CreateAuthenticator(connection, mergedDescription).Authenticate(connection, mergedDescription, cancellationToken);
```

Comment:
Sure, no problem. I initially chose this style for brevity's sake (and also because the line fit almost entirely within 120 characters, where one can reasonably expect "important" stuff to happen). But I have no issues introducing the interim variable.

Fixed!